misc/multistream-select/src/dialer_select.rs

```diff
@@ -58,371 +68,261 @@ where
+    FlushProtocol { io: MessageIO<R>, protocol: N },
+    AwaitProtocol { io: MessageIO<R>, protocol: N },
+    Done
 }
 
 impl<R, I> Future for DialerSelectSeq<R, I>
 where
     R: AsyncRead + AsyncWrite,
     I: Iterator,
-    I::Item: AsRef<[u8]> + Clone
+    I::Item: AsRef<[u8]>
 {
     type Item = (I::Item, Negotiated<R>);
-    type Error = ProtocolChoiceError;
+    type Error = NegotiationError;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         loop {
-            match mem::replace(&mut self.inner, DialerSelectSeqState::Undefined) {
-                DialerSelectSeqState::AwaitDialer { mut dialer_fut, mut protocols } => {
-                    let dialer = match dialer_fut.poll()? {
-                        Async::Ready(d) => d,
-                        Async::NotReady => {
-                            self.inner = DialerSelectSeqState::AwaitDialer { dialer_fut, protocols };
-                            return Ok(Async::NotReady)
-                        }
-                    };
-                    let proto_name = protocols.next().ok_or(ProtocolChoiceError::NoProtocolFound)?;
-                    self.inner = DialerSelectSeqState::NextProtocol {
-                        dialer,
-                        protocols,
-                        proto_name
+            match mem::replace(&mut self.state, SeqState::Done) {
+                SeqState::SendHeader { mut io } => {
+                    if io.start_send(Message::Header(Version::V1))?.is_not_ready() {
+                        self.state = SeqState::SendHeader { io };
+                        return Ok(Async::NotReady)
                     }
+                    let protocol = self.protocols.next().ok_or(NegotiationError::Failed)?;
+                    self.state = SeqState::SendProtocol { io, protocol };
                 }
-                DialerSelectSeqState::NextProtocol { mut dialer, protocols, proto_name } => {
-                    trace!("sending {:?}", proto_name.as_ref());
-                    let req = Request::Protocol { name: proto_name.clone() };
-                    match dialer.start_send(req)? {
-                        AsyncSink::Ready => {
-                            self.inner = DialerSelectSeqState::FlushProtocol {
-                                dialer,
-                                proto_name,
-                                protocols
-                            }
-                        }
-                        AsyncSink::NotReady(_) => {
-                            self.inner = DialerSelectSeqState::NextProtocol {
-                                dialer,
-                                protocols,
-                                proto_name
-                            };
-                            return Ok(Async::NotReady)
-                        }
+                SeqState::SendProtocol { mut io, protocol } => {
+                    let p = Protocol::try_from(protocol.as_ref())?;
```

Comment:
It is a little saddening that this PR reverses the direction of https://github.com/libp2p/rust-libp2p/pull/800.

Reply:
I was not completely unaware of that prior work. It is my understanding though (and you're probably aware) that `Bytes` / `BytesMut` [do not even allocate](https://docs.rs/bytes/0.4.12/bytes/struct.Bytes.html#inline-bytes) for `<= 31` bytes, which probably covers the majority of protocol names.

Together with the fact that this PR also generally reduces I/O roundtrips and the number of `send` / `recv` (sys)calls on the underlying transport, my hope is that there is overall certainly no severe performance regression (indeed, there should be an improvement, though I have no benchmarks to back up that claim, except for the fact that some tests started failing because of implicit assumptions on the "delay" caused by protocol negotiation over TCP transports). It may also be seen as a minor advantage for the implementation that cloning a `Protocol` is known to be cheap, whereas that is an implicit assumption for `N: Clone`.

I admit though that I'm generally not too concerned with performance on the level of counting allocations, unless there are benchmarks in place that call out major regressions.

As I noted in some inline comments, personally I would actually like to eventually replace the `N` type parameter in the API of this crate with the `Protocol` type, which this PR only does internally for now. Given that protocol names are actually supposed to be utf8-encoded strings, a related (vague) thought is that the internal representation of `Protocol` could possibly be changed later to `Cow<'static, str>` or something similar, which could then avoid allocations again also for larger, `'static` protocol names.

So overall, does that sound reasonable or what is your verdict on the PR? Do you generally approve of the changes or should more time be invested into analysing the performance and allocations, possibly by adding benchmarks (which could also be done as a follow-up, of course)?

Reply:
> So overall, does that sound reasonable or what is your verdict on the PR? Do you generally approve of the changes or should more time be invested into analysing the performance and allocations, possibly by adding benchmarks (which could also be done as a follow-up, of course)?

Overall I am in favour of this PR and I would not analyse the allocations any further. I have no doubts that due to the reduced number of round trips the overall performance is better. The creation of intermediate `Bytes` is just something I noticed while reading the diff.